File: Love2d Version/HopelessCatch_Source/world/daynightcycle.lua
-- DayNightCycle.lua - Manages day/night transitions and lighting
local DayNightCycle = {}

local timeOfDay = 0 -- 0 = midnight, 0.5 = noon, 1 = midnight again
local cycleSpeed = 0.0007 -- How fast time passes - Full cycle ~24 minutes (much slower)
local lightingCanvas
local ambientColor = {1, 1, 1, 1}
local isNight = false
local nightTransitionCooldown = 0
local NIGHT_TRANSITION_COOLDOWN_TIME = 2 -- seconds
local timeScale = 1.0 -- Always use recommended default timing
local currentDay = 1 -- Track which day we're on
local lastTimeOfDay = 0 -- Track previous time to detect day changes

-- Time phases
local phases = {
    {name = "Night", start = 0, r = 0.1, g = 0.1, b = 0.3},
    {name = "Dawn", start = 0.2, r = 0.8, g = 0.4, b = 0.2},
    {name = "Morning", start = 0.3, r = 1, g = 1, b = 0.9},
    {name = "Day", start = 0.4, r = 1, g = 1, b = 1},
    {name = "Evening", start = 0.7, r = 1, g = 0.7, b = 0.4},
    {name = "Dusk", start = 0.8, r = 0.4, g = 0.2, b = 0.6},
    {name = "Night", start = 0.9, r = 0.1, g = 0.1, b = 0.3}
}

function DayNightCycle:init()
    lightingCanvas = love.graphics.newCanvas(love.graphics.getDimensions())
    timeOfDay = 0.3 -- Start in morning
    self:updateAmbientColor()
    self.preloadedPhase = nil
end

function DayNightCycle:preloadNextPhase()
    -- Find next phase that we'll transition to
    local nextTime = timeOfDay + cycleSpeed * timeScale * 0.5
    if nextTime >= 1 then nextTime = nextTime - 1 end
    
    local nextPhase = nil
    for i, phase in ipairs(phases) do
        if nextTime >= phase.start and (i == #phases or nextTime < phases[i + 1].start) then
            nextPhase = phase
            break
        end
    end
    
    -- Preload if it's different from current
    if nextPhase and nextPhase ~= self.preloadedPhase then
        self.preloadedPhase = nextPhase
        collectgarbage("collect")  -- Help prevent stuttering
    end
end

function DayNightCycle:update(dt)
    -- Store previous time to detect day changes
    lastTimeOfDay = timeOfDay
    
    -- Advance time with smoother timing
    timeOfDay = (timeOfDay + cycleSpeed * timeScale * dt) % 1
    
    -- Check if we've crossed midnight (new day)
    if lastTimeOfDay > timeOfDay then
        currentDay = currentDay + 1
        self:onNewDay()
    end
    
    -- Preload any resources needed for next phase
    self:preloadNextPhase()
    
    -- Update ambient lighting with interpolation
    self:updateAmbientColor()
    
    -- Smooth night transition check
    local wasNight = isNight
    local newIsNight = timeOfDay < 0.3 or timeOfDay > 0.8
    
    -- Only update night state after cooldown
    if nightTransitionCooldown <= 0 then
        if wasNight ~= newIsNight then
            nightTransitionCooldown = NIGHT_TRANSITION_COOLDOWN_TIME
            isNight = newIsNight
        end
    else
        nightTransitionCooldown = nightTransitionCooldown - dt
    end
    
    -- Trigger night transition events instantly for fluid transitions
    if not wasNight and isNight then
        self:onNightFall()
    elseif wasNight and not isNight then
        self:onDayBreak()
    end
end

function DayNightCycle:updateAmbientColor()
    -- Find current phase
    local currentPhase = phases[1]
    for i = #phases, 1, -1 do
        if timeOfDay >= phases[i].start then
            currentPhase = phases[i]
            break
        end
    end
    
    -- Find next phase for interpolation
    local nextPhaseIndex = 1
    for i = 1, #phases do
        if phases[i].start > timeOfDay then
            nextPhaseIndex = i
            break
        end
    end
    
    local nextPhase = phases[nextPhaseIndex]
    
    -- Interpolate colors
    local progress = 0
    if nextPhase.start > currentPhase.start then
        progress = (timeOfDay - currentPhase.start) / (nextPhase.start - currentPhase.start)
    end
    
    progress = math.min(1, math.max(0, progress))
    
    ambientColor[1] = currentPhase.r + (nextPhase.r - currentPhase.r) * progress
    ambientColor[2] = currentPhase.g + (nextPhase.g - currentPhase.g) * progress
    ambientColor[3] = currentPhase.b + (nextPhase.b - currentPhase.b) * progress
    ambientColor[4] = 1
end

function DayNightCycle:beginLighting()
    love.graphics.push("all")
    love.graphics.setColor(ambientColor)
end

function DayNightCycle:endLighting()
    love.graphics.pop()
end

function DayNightCycle:drawTimeIndicator()
    -- Draw time indicator UI
    local x, baseY = 10, love.graphics.getHeight() - 140
    local boxWidth, boxHeight = 220, 120
    love.graphics.setColor(0, 0, 0, 0.7)
    love.graphics.rectangle("fill", x - 5, baseY - 5, boxWidth, boxHeight)

    love.graphics.setColor(1, 1, 1, 1)
    love.graphics.print("Day: " .. tostring(currentDay), x, baseY)
    love.graphics.print("Time: " .. self:getTimeString(), x, baseY + 20)
    love.graphics.print("Phase: " .. self:getCurrentPhase(), x, baseY + 40)
    love.graphics.print("Speed: " .. string.format("%.1fx", timeScale), x, baseY + 60)

    -- Day/night progress bar (moved to speed bar position, no speed bar)
    local barY = baseY + 85
    local barWidth = 120
    local barHeight = 10
    love.graphics.setColor(0.3, 0.3, 0.3, 1)
    love.graphics.rectangle("fill", x, barY, barWidth, barHeight)
    love.graphics.setColor(1, 1, 0.8, 1) -- Day color
    love.graphics.rectangle("fill", x + barWidth * 0.3, barY, barWidth * 0.5, barHeight)
    love.graphics.setColor(0.1, 0.1, 0.3, 1) -- Night color
    love.graphics.rectangle("fill", x, barY, barWidth * 0.3, barHeight)
    love.graphics.rectangle("fill", x + barWidth * 0.8, barY, barWidth * 0.2, barHeight)
    love.graphics.setColor(1, 0, 0, 1)
    love.graphics.rectangle("fill", x + barWidth * timeOfDay - 2, barY - 2, 4, barHeight + 4)
end

function DayNightCycle:getCurrentPhase()
    for i = #phases, 1, -1 do
        if timeOfDay >= phases[i].start then
            return phases[i].name
        end
    end
    return phases[1].name
end

function DayNightCycle:isNightTime()
    -- Direct time check instead of relying on isNight flag
    return timeOfDay < 0.3 or timeOfDay > 0.8
end

function DayNightCycle:getTimeOfDay()
    return timeOfDay
end

function DayNightCycle:setTime(time)
    timeOfDay = math.max(0, math.min(1, time))
    self:updateAmbientColor()
end

function DayNightCycle:onNightFall()
    print("Night falls... the horror begins...")
    -- Horror events disabled - no horrorevents module
    -- Audio disabled - no audiomanager module
end

function DayNightCycle:onDayBreak()
    print("Dawn breaks... safety returns...")
    -- Audio disabled - no audiomanager module
end

function DayNightCycle:onNewDay()
    print("A new day begins... Day " .. currentDay)
    -- Horror events disabled - no horrorevents module
end

-- Debug/Testing functions for time control
function DayNightCycle:setTimeScale(scale)
    timeScale = math.max(0, scale)
end

function DayNightCycle:getTimeScale()
    return timeScale
end

function DayNightCycle:pauseTime()
    timeScale = 0
end

function DayNightCycle:resumeTime()
    timeScale = 1.0
end

function DayNightCycle:fastForward(scale)
    timeScale = scale or 5.0
end

-- Get readable time format
function DayNightCycle:getTimeString()
    local hours = math.floor(timeOfDay * 24)
    local minutes = math.floor((timeOfDay * 24 * 60) % 60)
    return string.format("%02d:%02d", hours, minutes)
end

-- Get current day number
function DayNightCycle:getCurrentDay()
    return currentDay
end

-- Set day number (for save/load functionality)
function DayNightCycle:setCurrentDay(day)
    currentDay = day or 1
end

-- Advance to next day (for cabin sleep)
function DayNightCycle:advanceToNextDay()
    currentDay = currentDay + 1
    print("📅 Advanced to Day " .. currentDay)
end

return DayNightCycle

File: HTML_CSS_JS/src/daynightcycle.js
// DayNightCycle.js - port of world/daynightcycle.lua
(function(){
  const DayNightCycle = {
    // Time is 0-1 where 0.0 = midnight, 0.3 = morning, 0.5 = noon, 0.8 = evening, 1.0 = midnight
    time: 0.3,
    timeScale: 1.0,
    currentDay: 1,
    maxDays: 30,
    
    // Lighting state
    ambientLight: { r: 1, g: 1, b: 1, a: 1 },
    
    // Phase names
    phases: {
      NIGHT: 'night',
      DAWN: 'dawn',
      DAY: 'day',
      DUSK: 'dusk'
    },
    
    init() {
      this.time = 0.3; // Start at morning
      this.timeScale = 1.0;
      this.currentDay = 1;
      this.updateLighting();
    },
    
    update(dt) {
      if (window.menuActive) return;
      
      // Advance time
      this.time += (dt * this.timeScale) / 600; // 600 seconds = 10 minutes real time = 1 full day
      
      // Wrap around to next day
      if (this.time >= 1.0) {
        this.time = 0.0;
        this.currentDay = Math.min(this.currentDay + 1, this.maxDays);
        console.log('🌅 Day ' + this.currentDay + ' begins');
      }
      
      this.updateLighting();
    },
    
    updateLighting() {
      // Calculate lighting based on time of day
      // 0.0-0.2 = night (dark)
      // 0.2-0.35 = dawn (brightening)
      // 0.35-0.65 = day (bright)
      // 0.65-0.8 = dusk (dimming)
      // 0.8-1.0 = night (dark)
      
      let r = 1, g = 1, b = 1, a = 1;
      
      if (this.time < 0.2) {
        // Night: very dark blue
        const t = this.time / 0.2;
        r = 0.3 + t * 0.2;
        g = 0.3 + t * 0.3;
        b = 0.5 + t * 0.2;
        a = 0.6;
      } else if (this.time < 0.35) {
        // Dawn: brightening
        const t = (this.time - 0.2) / 0.15;
        r = 0.5 + t * 0.5;
        g = 0.6 + t * 0.4;
        b = 0.7 - t * 0.2;
        a = 0.8 + t * 0.2;
      } else if (this.time < 0.65) {
        // Day: bright
        r = 1.0;
        g = 1.0;
        b = 0.5;
        a = 1.0;
      } else if (this.time < 0.8) {
        // Dusk: dimming
        const t = (this.time - 0.65) / 0.15;
        r = 1.0 - t * 0.5;
        g = 1.0 - t * 0.4;
        b = 0.5 - t * 0.2;
        a = 1.0 - t * 0.2;
      } else {
        // Night: dark
        const t = (this.time - 0.8) / 0.2;
        r = 0.5 - t * 0.2;
        g = 0.6 - t * 0.3;
        b = 0.3 + t * 0.2;
        a = 0.8 - t * 0.2;
      }
      
      this.ambientLight = { r, g, b, a };
    },
    
    getTime() {
      return this.time;
    },
    
    setTime(newTime) {
      this.time = Math.max(0, Math.min(1, newTime));
      this.updateLighting();
    },
    
    getCurrentDay() {
      return this.currentDay;
    },
    
    getDayCount() {
      return this.currentDay;
    },
    
    setCurrentDay(day) {
      this.currentDay = Math.max(1, Math.min(day, this.maxDays));
    },
    
    getPhase() {
      if (this.time < 0.2 || this.time >= 0.8) return this.phases.NIGHT;
      if (this.time < 0.35) return this.phases.DAWN;
      if (this.time < 0.65) return this.phases.DAY;
      return this.phases.DUSK;
    },
    
    isNight() {
      return this.time < 0.2 || this.time >= 0.8;
    },
    
    isDay() {
      return this.time >= 0.35 && this.time < 0.65;
    },
    
    isDawn() {
      return this.time >= 0.2 && this.time < 0.35;
    },
    
    isDusk() {
      return this.time >= 0.65 && this.time < 0.8;
    },
    
    getAmbientLight() {
      return this.ambientLight;
    },
    
    setTimeScale(scale) {
      this.timeScale = Math.max(0, scale);
    },
    
    getTimeScale() {
      return this.timeScale;
    },
    
    advanceToNextDay() {
      this.time = 0.3; // Morning
      this.currentDay = Math.min(this.currentDay + 1, this.maxDays);
      this.updateLighting();
      console.log('🌅 Advanced to Day ' + this.currentDay);
    },
    
    // Draw time indicator on HUD
    drawTimeIndicator() {
      const canvas = document.getElementById('game-canvas');
      if (!canvas) return;
      
      const ctx = canvas.getContext('2d');
      if (!ctx) return;
      
      // Draw time display in top-left
      ctx.save();
      ctx.fillStyle = 'rgba(0, 0, 0, 0.5)';
      ctx.fillRect(10, 10, 120, 50);
      
      ctx.fillStyle = '#fff';
      ctx.font = 'bold 12px Arial';
      ctx.fillText('Day ' + this.currentDay, 15, 28);
      
      const hours = Math.floor(this.time * 24);
      const minutes = Math.floor((this.time * 24 - hours) * 60);
      const timeStr = String(hours).padStart(2, '0') + ':' + String(minutes).padStart(2, '0');
      ctx.fillText(timeStr, 15, 43);
      
      ctx.restore();
    },
    
    // Apply lighting to canvas (called before drawing game objects)
    beginLighting() {
      const canvas = document.getElementById('game-canvas');
      if (!canvas) return;
      
      const ctx = canvas.getContext('2d');
      if (!ctx) return;
      
      const light = this.ambientLight;
      ctx.save();
      ctx.globalAlpha = light.a;
      ctx.fillStyle = 'rgb(' + Math.floor(light.r * 255) + ',' + Math.floor(light.g * 255) + ',' + Math.floor(light.b * 255) + ')';
      ctx.fillRect(0, 0, canvas.width, canvas.height);
    },
    
    // End lighting effects
    endLighting() {
      const canvas = document.getElementById('game-canvas');
      if (!canvas) return;
      
      const ctx = canvas.getContext('2d');
      if (!ctx) return;
      
      ctx.restore();
    }
  };
  
  if (typeof window !== 'undefined') window.DayNightCycle = window.DayNightCycle || DayNightCycle;
  if (typeof module !== 'undefined') module.exports = DayNightCycle;
})();
